Add documentation on using virtual environments for testing in napari based on 2022-01 workshop by Talley Lambert (#50)
napari/docs: docs/plugins/virtual_environment_docs/5-survey.md

- [How to find the different signals or slots?](#how-to-find-the-different-signals-or-slots)  
- [How do you create or delete a point or shape layer?](#how-do-you-create-or-delete-a-point-or-shape-layer)  
- [How do you avoid github tests failing?](#how-do-you-avoid-github-tests-failing)  
- [How do you make a process cancellable](#how-do-you-make-a-process-cancellable)  
- [Are there testing environments in napari?](#are-there-testing-environments-in-napari)  
- [Introduction to npe2? Migrating to new plug-in architecture?](#introduction-to-npe2-migrating-to-new-plug-in-architecture)  
- [What is the optimal setup to quickly iterate in widget development?](#what-is-the-optimal-setup-to-quickly-iterate-in-widget-development)  
  
## What are the best practices to test a plugin with multiple sequential steps?
e.g. Is it ok to rely on the "results" of a test to run the next test or should they all be fully independent?  

Answer:  
* Ideally, aim for unit testing.  
* Tests should not pass/fail together.  
* Use fixtures to provide a test with inputs, even if you have to make them up.  
  https://docs.pytest.org/en/6.2.x/fixture.html 

> * Use mocks (mock-ups) to assert that specific calls are made, without necessarily caring about what happens after that call is made.  

Reviewer: For both of these the link should be a hyperlink. One on the word `fixtures` and one on the word `mocks`

  https://docs.python.org/3/library/unittest.mock.html 

*This is definitely an art form. It takes time. Be patient.*

## How do you test widgets, the napari viewer, graphical user interfaces, and Qt in general?
Answer:
* Try not to!
* You should generally trust that a button click (for example) will call your callback and focus on testing that your callback does what it's supposed to do given that it gets called following some UI interaction.
* However: If you have a scenario where you are actually creating a complicated widget directly in Qt, see `pytest-qt` for lots of tips, specifically `qtbot`.

>     - pytest-qt - [https://pytest-qt.readthedocs.io/en/latest/intro.html](https://pytest-qt.readthedocs.io/en/latest/intro.html)

Reviewer: ```suggestion
    - [pytest-qt](https://pytest-qt.readthedocs.io/en/latest/intro.html)
```

>     - qtbot - [https://pytest-qt.readthedocs.io/en/latest/reference.html?highlight=qtbot#module-pytestqt.qtbot](https://pytest-qt.readthedocs.io/en/latest/reference.html?highlight=qtbot#module-pytestqt.qtbot)

Reviewer: ```suggestion
    - [qtbot](https://pytest-qt.readthedocs.io/en/latest/reference.html?highlight=qtbot#module-pytestqt.qtbot)
```

* Oftentimes, this comes down to knowing and/or learning the Qt API really well.  

## How to find the different signals or slots?
Question: How can we find the different signals/slots we can connect callbacks to as the user interacts with the core napari interface e.g. creating/editing/deleting a `points` or `shapes` layer?

Answer: 
[https://napari.org/guides/stable/events_reference.html](https://napari.org/stable/guides/events_reference.html)  
Granted, this is a work in progress. 

> ## How do you create or delete a point or shape layer?  

Reviewer: This is still part of the same question about different signals and slots.

```suggestion
```

Author: Removed & incorporated into the signals & slots question.

For example, these events are emitted when the user interacts with the layer list: 
```console  
    Viewer.layers.events.inserted  
    Viewer.layers.events.removed  
    Viewer.layers.events.moved  
    Viewer.layers.events.changed  
    Viewer.layers.events.reordered  
```    

> Getting an event when the user is editing the data inside a point or shape layer (outside of the GUI interface) is complicated, because the user will be directly editing the native array object.

Reviewer: ```suggestion
Getting an event when the user is editing the data inside a `points` or `shapes` layer (outside of the GUI interface) is complicated, because the user will be directly editing the native array object.
```

## How do you avoid github tests failing?  
Answer:  
* First make sure all your tests are passing locally.  
* After that, it's complicated. More background or context is needed to answer this question.  
  
## How do you make a process cancellable? 
Question: How do you make a process cancellable to interrupt a method that is running in a for loop, for example?  

Answer:  
* In single-threaded python, use `Ctrl-C`  

> * In multithreaded python, there are many different patterns. Consider using a generator-based thread worker:  

Reviewer: Should be a hyperlink on `generator-based thread worker`

  https://napari.org/stable/guides/threading.html#generators-for-the-win  

## Are there testing environments in napari?
Answer: Napari does not create or otherwise manage environments.  
  

> ## Introduction to npe2? Migrating to new plug-in architecture?     

Reviewer: ```suggestion
## Introduction to npe2? Migrating to new plugin architecture?     
```

# Survey/Q&A

This guide contains questions that were submitted to our survey on testing.  

## This guide covers:   
* [Questions](#questions)  
    - [Testing plugins](#testing-plugins)  
    - [Testing widgets](#testing-widgets)  
    - [Better signal or slot documentation or functionality](#better-signal-or-slot-documentation-or-functionality)  
    - [Creating or deleting a point or shape layer](#creating-or-deleting-a-point-or-shape-layer)  
    - [Github tests failing](#github-tests-failing)  
    - [Cancelling a method](#cancelling-a-method)  
    - [Testing environments](#testing-environments)  
* [Topics](#topics)  
    - [npe2](#npe2)  
    - [Widget development](#widget-development)
  

> ## Questions

Reviewer: I think we should get rid of `##Questions` and `##Topics` as headings. The distinction isn't really there.

Author: Please see what you think now.

> ### Testing plugins    

Reviewer: These headings should just be the questions themselves. This would allow them to show in the sidebar, once we bump these headings to `##` after removing the `Questions` and `Topics`.

Author: Please see what you think now.

Question: What are the best practices to test a plugin with multiple sequential steps?  
e.g. Is it ok to rely on the "results" of a test to run the next test or should they all be fully independent?  

> Answer:  

Reviewer: Once we make the headings themselves questions, the answer can just be the content under that heading and we can get rid of this.

Author: I fixed this.

  https://docs.python.org/3/library/unittest.mock.html 

*This is definitely an art form. It takes time. Be patient.*

### Testing widgets
Question: How do you test widgets, the napari viewer, and graphical user interfaces? How do you test Qt in general?

Answer:
* Try not to!
* You should generally trust that a button click (for example) will call your callback and focus on testing that your callback does what it's supposed to do given that it gets called following some UI interaction.
* However: If you have a scenario where you are actually creating a complicated widget directly in Qt, see `pytest-qt` for lots of tips, specifically `qtbot`.
    - pytest-qt - [https://pytest-qt.readthedocs.io/en/latest/intro.html](https://pytest-qt.readthedocs.io/en/latest/intro.html)
    - qtbot - [https://pytest-qt.readthedocs.io/en/latest/reference.html?highlight=qtbot#module-pytestqt.qtbot](https://pytest-qt.readthedocs.io/en/latest/reference.html?highlight=qtbot#module-pytestqt.qtbot)
* Oftentimes, this comes down to knowing and/or learning the Qt API really well.  

### Better signal or slot documentation or functionality

> Question: How to create better signal/slot documentation/functionality as the user interacts with the core napari interface? For example, creating/editing/deleting a point or shape layer?

Reviewer: The question here is suuper vague and tbh I don't really understand what it's asking. I would tweak this to match what we're actually answering.

```suggestion
Question: How can we find the different signals/slots we can connect callbacks to as the user interacts with the core napari interface e.g. creating/editing/deleting a `points` or `shapes` layer?
```

Answer: 
[https://napari.org/guides/stable/events_reference.html](https://napari.org/stable/guides/events_reference.html)  
Granted, this is a work in progress. 

> ### Creating or deleting a point or shape layer:  

Reviewer: ```suggestion
For example, these events are emitted when the user interacts with the layer list:  
```

>     Viewer.layers.events.inserted  

Reviewer: These should all be in a code block

Author: Done.

    Viewer.layers.events.removed  
    Viewer.layers.events.moved  
    Viewer.layers.events.changed  
    Viewer.layers.events.reordered  

> Editing a point or shape layer may be complicated, because it may be a pure NumPy object. 

Reviewer: ```suggestion
Getting an event when the user is editing the data inside a point or shape layer (outside of the GUI interface) is complicated, because the user will be directly editing the native array object.
```

> # Tools and tips when building napari plugins: Survey 

Reviewer: ```suggestion
# 5: Survey
```

Author: Done.

# Survey/Q&A

This guide contains questions that were submitted to our survey on testing.  

## This guide covers:   
- [What are the best practices to test a plugin with multiple sequential steps?](#what-are-the-best-practices-to-test-a-plugin-with-multiple-sequential-steps)  
- [How do you test widgets, the napari viewer, graphical user interfaces, and Qt in general?](#how-do-you-test-widgets-the-napari-viewer-graphical-user-interfaces-and-qt-in-general)  
- [How to find the different signals or slots?](#how-to-find-the-different-signals-or-slots)  
- [How do you avoid github tests failing?](#how-do-you-avoid-github-tests-failing)  
- [How do you make a process cancellable](#how-do-you-make-a-process-cancellable)  
- [Are there testing environments in napari?](#are-there-testing-environments-in-napari)  

> - [Introduction to npe2? Migrating to new plug-in architecture?](#introduction-to-npe2-migrating-to-new-plug-in-architecture)  

Reviewer 2: ```suggestion
- [Introduction to npe2? Migrating to new plug-in architecture?](#introduction-to-npe2-migrating-to-new-plugin-architecture)  
```

Answer:  
* Ideally, aim for unit testing.  
* Tests should not pass/fail together.  
* Use [fixtures](https://docs.pytest.org/en/6.2.x/fixture.html) to provide a test with inputs, even if you have to make them up.  
* Use [mocks (mock-ups)](https://docs.python.org/3/library/unittest.mock.html) to assert that specific calls are made, without necessarily caring about what happens after that call is made.  

*This is definitely an art form. It takes time. Be patient.*

## How do you test widgets, the napari viewer, graphical user interfaces, and Qt in general?
Answer:
* Try not to!
* You should generally trust that a button click (for example) will call your callback and focus on testing that your callback does what it's supposed to do given that it gets called following some UI interaction.
* However: If you have a scenario where you are actually creating a complicated widget directly in Qt, see `pytest-qt` for lots of tips, specifically `qtbot`.
    - [pytest-qt](https://pytest-qt.readthedocs.io/en/latest/intro.html)
    - [qtbot](https://pytest-qt.readthedocs.io/en/latest/reference.html?highlight=qtbot#module-pytestqt.qtbot)

> * Oftentimes, this comes down to knowing and/or learning the Qt API really well.  

Reviewer: We should link to the testing workshop here!

> * Please see also the [In-depth guide to plugin testing](../../../../docs/plugins/testing_workshop_docs).

Reviewer 2: ```suggestion
* Please see also the [In-depth guide to plugin testing](../testing_workshop_docs/index.md).
```
Need to link to the markdown, I think this should make it work!

Answer: 
[https://napari.org/guides/stable/events_reference.html](https://napari.org/stable/guides/events_reference.html)  

Granted, this is a work in progress. 

For example, these events are emitted when the user interacts with the layer list: 
```console  
    Viewer.layers.events.inserted  
    Viewer.layers.events.removed  
    Viewer.layers.events.moved  
    Viewer.layers.events.changed  
    Viewer.layers.events.reordered  
```    

Getting an event when the user is editing the data inside a `points` or `shapes` layer (outside of the GUI interface) is complicated, because the user will be directly editing the native array object.

## How do you avoid github tests failing?  
Answer:  
* First make sure all your tests are passing locally.  
* After that, it's complicated. More background or context is needed to answer this question.  
  
## How do you make a process cancellable? 
Question: How do you make a process cancellable to interrupt a method that is running in a for loop, for example?  

Answer:  
* In single-threaded python, use `Ctrl-C`  

> * In multithreaded python, there are many different patterns. Consider using a [generator-based thread worker].(https://napari.org/stable/guides/threading.html#generators-for-the-win)    

Reviewer 2: ```suggestion
* In multithreaded python, there are many different patterns. Consider using a [generator-based thread worker](https://napari.org/stable/guides/threading.html#generators-for-the-win).    
```

## Are there testing environments in napari?
Answer: Napari does not create or otherwise manage environments.  
  
## Introduction to npe2? Migrating to new plugin architecture?     
Answer:  
* The primary difference is in how plugins are discovered:  
    - npe1 used decorators, requiring module import.  
    - npe2 uses static manifests (`napari.yaml`), describing contributions without requiring import.  

>     - Your first plugin: https://napari.org/stable/plugins/first_plugin.html  

Reviewer 2: ```suggestion
    - See also the [Your First Plugin tutorial](https://napari.org/stable/plugins/first_plugin.html)  
```

> * Contributions Reference: https://napari.org/stable/plugins/contributions.html  
> * Guides for each type of contribution: https://napari.org/stable/plugins/guides.html  
> * Migration guide: https://napari.org/stable/plugins/npe2_migration_guide.html  

Reviewer 2: ```suggestion
Additional resources:
* [Contributions Reference](https://napari.org/stable/plugins/contributions.html)  
* [Guides for each type of contribution](https://napari.org/stable/plugins/guides.html)  
* [Migration guide](https://napari.org/stable/plugins/npe2_migration_guide.html)  
```

These bare links are clickable in GitHub as markdown, but not in the rendered html pages. This should make them work properly.

  
## What is the optimal setup to quickly iterate in widget development?
Answer:   
* Create a script that will start napari and load your widget without any UI interaction.  
* Don't test as a plugin. Start by directly calling `viewer.window.add_dock_widget` with a manually created widget.  

> * Familiarize yourself with auto-reload features.  
>   https://ipython.readthedocs.io/en/stable/config/extensions/autoreload.html   

Reviewer 2: ```suggestion
* Familiarize yourself with the [IPython auto-reload features](https://ipython.readthedocs.io/en/stable/config/extensions/autoreload.html).   
```
Same as above

> * Consider using watchmedo from watchdog:  
>   https://github.com/gorakhargosh/watchdog  

Reviewer 2: ```suggestion
* Consider using `watchmedo` from [watchdog](https://github.com/gorakhargosh/watchdog).
```
